src/main/java/org/testng/DependencyMap.java

```diff
@@ -82,13 +83,30 @@ private static boolean hasInstance(
       ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
     Object baseInstance = baseClassMethod.getInstance();
     Object derivedInstance = derivedClassMethod.getInstance();
-    return derivedInstance != null || baseInstance != null;
+    boolean result = derivedInstance != null || baseInstance != null;
+    boolean params = null != baseClassMethod.getFactoryMethodParamsInfo() && null != derivedClassMethod.getFactoryMethodParamsInfo().getParameters();
+
+    if (result && params && RuntimeBehavior.enforceThreadAffinity()) {
+      return hasSameParameters(baseClassMethod, derivedClassMethod);
+    }
+    return result;
+  }
+
+  private static boolean hasSameParameters(
+      ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
+    return baseClassMethod.getFactoryMethodParamsInfo().getParameters()[0]
+        .equals(derivedClassMethod.getFactoryMethodParamsInfo().getParameters()[0]);
   }
 
   private static boolean isSameInstance(
       ITestNGMethod baseClassMethod, ITestNGMethod derivedClassMethod) {
     Object baseInstance = baseClassMethod.getInstance();
     Object derivedInstance = derivedClassMethod.getInstance();
+    boolean result = derivedInstance != null && baseInstance != null;
```

Comment:
I propose:

```java
if (derivedInstance != null && baseInstance != null) {
  if (null != baseClassMethod.getFactoryMethodParamsInfo() && RuntimeBehavior.enforceThreadAffinity()) {
    return baseInstance.getClass().isAssignableFrom(derivedInstance.getClass()) &&
          hasSameParameters(baseClassMethod, derivedClassMethod);  
  } else {
    return baseInstance.getClass().isAssignableFrom(derivedInstance.getClass());
  }
} else {
  return false;
}
```

It highlights the difference between the 2 options.
Quick return could be used too.

Reply:
Fixed